### PR TITLE
Add GLM-ASR multimodal support 

### PR DESCRIPTION
**Pull request by baonudesifeizhai**
## Purpose
Support that model :https://huggingface.co/zai-org/GLM-ASR-Nano-2512 
## Test Result
for download and convert test audio:
https://paste.ubuntu.com/p/K7S9Thmvfg/

Start server:
```
export VLLM_DISABLE_TORCH_COMPILE=1
export TORCH_COMPILE=0

python -m vllm.entrypoints.openai.api_server \
  --model zai-org/GLM-ASR-Nano-2512 \
  --port 8000 \
  --dtype bfloat16 \
  --allowed-local-media-path /root/vllm/peoples_speech_100
```

 bench_glm_asr_e2e.py https://paste.ubuntu.com/p/t658x4W2fJ/ 
python bench_glm_asr_e2e.py --dir /root/vllm/peoples_speech_100 --limit 100 --concurrency 1 --warmup 1
```
  === SUMMARY ===
 OK=100 FAIL=0
 Wall time: 43.08s Latency(s): mean=0.43 p50=0.42 p90=0.50 p99=0.63 
RTF: mean=0.03 p50=0.03 p90=0.04 p99=0.04 Total audio: 1449.06s => Throughput=33.64 audio-sec/s Saved: /root/vllm/peoples_speech_100/bench_100_c1.jsonl Saved: /root/vllm/peoples_speech_100/transcripts_100_c1.tsv

```
python bench_glm_asr_e2e.py --dir /root/vllm/peoples_speech_100 --limit 100 --concurrency 4 --warmup 1
```
=== SUMMARY === OK=100 FAIL=0 Wall time: 9.81s Latency(s): mean=0.39 p50=0.38 p90=0.48 p99=0.55 RTF: mean=0.03 p50=0.03 p90=0.03 p99=0.04 Total audio: 1449.06s => Throughput=147.64 audio-sec/s Saved: /root/vllm/peoples_speech_100/bench_100_c4.jsonl Saved: /root/vllm/peoples_speech_100/transcripts_100_c4.tsv (.venv) root@d50ebd3703cf:~/vllm

```
python bench_glm_asr_e2e.py --dir /root/vllm/peoples_speech_100 --limit 100 --concurrency 8 --warmup 1
```
`
=== SUMMARY === OK=100 FAIL=0 Wall time: 5.36s Latency(s): mean=0.41 p50=0.41 p90=0.53 p99=0.59 RTF: mean=0.03 p50=0.03 p90=0.04 p99=0.04 Total audio: 1449.06s => Throughput=270.24 audio-sec/s Saved: /root/vllm/peoples_speech_100/bench_100_c8.jsonl Saved: /root/vllm/peoples_speech_100/transcripts_100_c8.tsv (.venv) root@d50ebd3703cf:~/vllm#
`
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by jeejeelee** [comment]
Thank you for the contribution!  Regarding the new MM model, we also need to:

- Add model information in [list-of-multimodal-language-models](https://github.com/vllm-project/vllm/blob/v0.14.0rc0/docs/models/supported_models.md#list-of-multimodal-language-models)
- Register this model in [registry](https://github.com/vllm-project/vllm/blob/v0.14.0rc0/tests/models/registry.py#L582)
- Add a usage example in [audio_language.py](https://github.com/vllm-project/vllm/blob/v0.14.0rc0/examples/offline_inference/audio_language.py)

File: vllm/model_executor/models/glmasr.py [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+        TensorShape("num_chunks", "nmb", "chunk_length", dynamic_dims={"chunk_length"}),
+    ]
+    feature_attention_mask: Annotated[
+        torch.Tensor,
+        TensorShape("num_chunks", "chunk_length", dynamic_dims={"chunk_length"}),
+    ]
+    chunk_counts: Annotated[
+        torch.Tensor,
+        TensorShape("num_audios"),
+    ]
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/glmasr.py:97]: Please address this

File: vllm/model_executor/models/glmasr.py [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+from collections.abc import Iterable, Mapping, Sequence
+from typing import Annotated, Any, Literal, TypeAlias
+
+import torch
+import torch.nn as nn
+from transformers import BatchFeature
+from transformers.models.glmasr import GlmAsrConfig, GlmAsrEncoder, GlmAsrProcessor
+from transformers.models.whisper import WhisperFeatureExtractor
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py:12]: You can use the Whisper model from vLLM

Comment (baonudesifeizhai) [reply vllm/model_executor/models/glmasr.py]: We’re using WhisperFeatureExtractor for preprocessing, but GLM‑ASR has a different encoder/projector stack, so we can’t swap in vLLM’s Whisper model directly. 
https://github.com/sgl-project/sglang/pull/15570/changes  also same  as sglang...

File: vllm/model_executor/models/glmasr_utils.py
Comment (Isotr0py) [inline vllm/model_executor/models/glmasr_utils.py:1]: I think this is too model-specific, we can move it into model implementation `glmasr.py` or `vllm/model_executor/models/glmasr_utils.py`.

File: vllm/model_executor/models/glmasr.py [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+            return DictEmbeddingItems(
+                data,
+                modality="audio",
+                required_fields={"audio_embeds"},
+                fields_factory=_glmasr_field_config,
+            )
+        return super()._parse_audio_data(data)
+
+
+class GlmAsrMultiModalProcessor(BaseMultiModalProcessor[GlmAsrProcessingInfo]):
```
Comment (Isotr0py) [inline vllm/model_executor/models/glmasr.py]: BTW, seems GLMASR use similar processor like AudioFlamingo3, perhpas we can reuse AudioFlamingo3's multimodal processor?
https://github.com/huggingface/transformers/blob/a7f29523361b2cc12e51c1f5133d95f122f6f45c/src/transformers/models/glmasr/modular_glmasr.py#L53-L76
https://github.com/vllm-project/vllm/blob/96142f209453a381fcaf9d9d010bbf8711119a77/vllm/model_executor/models/audioflamingo3.py#L277-L279

File: vllm/model_executor/models/glmasr.py [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+        super().__init__()
+        self.linear_1 = nn.Linear(
+            config.audio_config.intermediate_size,
+            config.text_config.hidden_size * 2,
+        )
+        self.act = get_act_fn(config.projector_hidden_act)
+        self.linear_2 = nn.Linear(
+            config.text_config.hidden_size * 2,
+            config.text_config.hidden_size,
+        )
```
Comment (Isotr0py) [inline vllm/model_executor/models/glmasr.py]: Use `ColumnParallelLinear` and `RowParallelLinear` here.

File: vllm/model_executor/models/glmasr.py (hunk 2 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/glmasr.py:12]: Nvm it is only used in preprocessing, so it's fine

**Comment by baonudesifeizhai** [comment]
will try AudioFlamingo3MultiModalDataParser later

File: docs/models/supported_models.md [hunk 5]
```diff
@@ -773,6 +773,7 @@ Speech2Text models trained specifically for Automatic Speech Recognition.
 
 | Architecture | Models | Example HF Models | [LoRA](../features/lora.md) | [PP](../serving/parallelism_scaling.md) |
 |--------------|--------|-------------------|----------------------|---------------------------|
+| `GlmAsrForConditionalGeneration` | GLM-ASR | `zai-org/GLM-ASR-Nano-2512` | ✅︎ | ✅︎ |
```
Comment (DarkLight1337) [inline docs/models/supported_models.md]: While you're here, can you move Gemma3n and Granite Speech so the table is sorted in alphabetical order?

File: vllm/model_executor/models/glmasr.py [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+
+        return {
+            "audio": self._get_dummy_audios(
+                length=audio_len, num_audios=num_audios, overrides=audio_overrides
+            )
+        }
+
+
+# Reuse field config from AudioFlamingo3
+_glmasr_field_config = _audioflamingo3_field_config
```
Comment (DarkLight1337) [inline vllm/model_executor/models/glmasr.py]: You can use `import ... as ...` statement to do the renaming at the top of the file

File: docs/models/supported_models.md (hunk 5 above)
Comment (DarkLight1337) [reply docs/models/supported_models.md]: Also, the model should implement `SupportsTranscription` if you want it to be used with Transcriptions API

**Comment by baonudesifeizhai** [comment]
```
(.venv) root@16c22feb96e4:~/vllm# for i in 03 04 05; do
  curl http://localhost:8000/v1/audio/transcriptions \
    -H "Authorization: Bearer dummy" \
    -F "model=zai-org/GLM-ASR-Nano-2512" \
    -F "file=@/root/vllm/peoples_speech_100/07282016HFUUforum_SLASH_07-28-2016_HFUUforum_DOT_mp3_000${i}.flac.wav" \
    -F "response_format=text"
  echo -e "\n---\n"
done
{"text":"And we produce last year we produced twenty one thousand pounds of food for the community on two thousand square feet. So unless we were that efficient to produce that much food.","usage":{"type":"duration","seconds":15}}
---

{"text":"Commons here in the spirit of being able to grow our food here locally. So, we started as an organization in actually in two thousand nine, but really in two thousand ten, March of two thousand ten, we had our first meeting.","usage":{"type":"duration","seconds":15}}
---

{"text":"Maui, and now we have ten chapters across the islands, and about nine hundred members. So we are growing exponentially, and we're fortunate to know that we have not only.","usage":{"type":"duration","seconds":15}}
---

(.venv) root@16c22feb96e4:~/vllm#

```

support transcript api right now!

**Review by DarkLight1337 — APPROVED** [review]
LGTM now, thanks!

**Comment by DarkLight1337** [comment]
PTAL at the failing processor test, you might need to set `trust_remote_code=True` in the test registry

**Comment by baonudesifeizhai** [comment]
image
 god...

**Comment by baonudesifeizhai** [comment]
Use a patch function to restrict the number of audio inputs to one, aligning with real-world usage (the Transcription API processes one audio file per request).  
tests/models/test_initialization.py::test_can_initialize_large_subset[GlmAsrForConditionalGeneration] PASSED     [100%]
tests/models/multimodal/processing/test_tensor_schema.py::test_model_tensor_schema[zai-org/GLM-ASR-Nano-2512] PASSED [100%]

File: tests/models/registry.py [hunk 7]
```diff
@@ -647,6 +647,9 @@ def check_available_online(
     "FuyuForCausalLM": _HfExamplesInfo("adept/fuyu-8b"),
     "Gemma3ForConditionalGeneration": _HfExamplesInfo("google/gemma-3-4b-it"),
     "Gemma3nForConditionalGeneration": _HfExamplesInfo("google/gemma-3n-E2B-it"),
+    "GlmAsrForConditionalGeneration": _HfExamplesInfo(
+        "zai-org/GLM-ASR-Nano-2512", trust_remote_code=True
```
Comment (DarkLight1337) [inline tests/models/registry.py]: Is Transformers v5 required to use this model? If so you should also add `min_transformers_version`
